website_cms/controllers/main.py

```diff
@@ -0,0 +1,159 @@
+
+    def render(self, main_object, **kw):
+        """Retrieve parameters for rendering and render view template."""
+        return request.website.render(
+            self.get_template(main_object, **kw),
+            self.get_render_values(main_object, **kw),
+        )
+
+
+# `secure_model` is our converter that checks security
+# see `website.security.mixin`.
+PAGE_VIEW_ROUTES = [
+    '/cms/<secure_model("cms.page"):main_object>',
+    '/cms/<path:path>/<secure_model("cms.page"):main_object>',
+    '/cms/<secure_model("cms.page"):main_object>/page/<int:page>',
+    '/cms/<path:path>/<secure_model("cms.page"):main_object>/page/<int:page>',
+    '/cms/<secure_model("cms.page"):main_object>\
+        /media/<model("cms.media.category"):media_categ>',
+    '/cms/<secure_model("cms.page"):main_object>\
+        /media/<model("cms.media.category"):media_categ>/page/<int:page>',
+    '/cms/<path:path>/<secure_model("cms.page"):main_object>\
+        /media/<model("cms.media.category"):media_categ>',
+    '/cms/<path:path>/<secure_model("cms.page"):main_object>\
+        /media/<model("cms.media.category"):media_categ>/page/<int:page>',
+]
+
+
+class PageViewController(http.Controller, ContextAwareMixin):
+    """CMS page view controller."""
+
+    template = 'website_cms.page_default'
+
+    @http.route(PAGE_VIEW_ROUTES, type='http', auth='public', website=True)
+    def view_page(self, main_object, **kw):
+        """Handle a `page` route.
+
+        Many optional arguments come from `kw` based on routing match above.
+        """
+        site = request.website
+        # check published
+        # This is weird since it should be done by `website` module itself.
+        # Alternatively we can put this in our `secure model` route handler.
+        if not site.is_publisher() and not main_object.website_published:
+            raise NotFound
+
+        if main_object.has_redirect():
+            data = main_object.get_redirect_data()
+            redirect = werkzeug.utils.redirect(data.url, data.status)
+            return redirect
+        if 'edit_translations' in kw:
+            # for some reasons here we get an empty string
+            # as value, and this breaks translation editor initialization :(
+            kw['edit_translations'] = True
+
+        # handle translations switch
+        if site and 'edit_translations' not in kw \
+                and not site.default_lang_code == request.lang \
+                and not site.is_publisher():
+            # check if there's any translation for current page in request lang
+            if request.lang not in main_object.get_translations():
+                raise NotFound
+        return self.render(main_object, **kw)
+
+
+class TagsController(http.Controller):
+    """CMS tags controller."""
+
+    @http.route('/cms/get_tags', type='http',
+                auth="public", methods=['GET'], website=True)
+    def tags_search(self, q='', l=25, **post):
```

Comment:
Note that using JSON RPC will disallow the public user from using this application. This may be a hinderance to some usage, thus I would say we should probably leave the controller until I finish OCA/web#402